vms/evm/metrics/prometheus/prometheus_test.go

```diff
@@ -0,0 +1,168 @@
+
+	counterFloat64 := metrics.NewCounterFloat64()
+	counterFloat64.Inc(1.1)
+	register(t, "test/counter_float64", counterFloat64)
+
+	gauge := metrics.NewGauge()
+	gauge.Update(23456)
+	register(t, "test/gauge", gauge)
+
+	gaugeFloat64 := metrics.NewGaugeFloat64()
+	gaugeFloat64.Update(34567.89)
+	register(t, "test/gauge_float64", gaugeFloat64)
+
+	gaugeInfo := metrics.NewGaugeInfo()
+	gaugeInfo.Update(metrics.GaugeInfoValue{"key": "value"})
+	register(t, "test/gauge_info", gaugeInfo) // skipped
+
+	sample := metrics.NewUniformSample(1028)
+	histogram := metrics.NewHistogram(sample)
+	register(t, "test/histogram", histogram)
+
+	meter := metrics.NewMeter()
+	t.Cleanup(meter.Stop)
+	meter.Mark(9999999)
+	register(t, "test/meter", meter)
+
+	timer := metrics.NewTimer()
+	t.Cleanup(timer.Stop)
+	timer.Update(20 * time.Millisecond)
+	timer.Update(21 * time.Millisecond)
+	timer.Update(22 * time.Millisecond)
+	timer.Update(120 * time.Millisecond)
+	timer.Update(23 * time.Millisecond)
+	timer.Update(24 * time.Millisecond)
+	register(t, "test/timer", timer)
+
+	resettingTimer := metrics.NewResettingTimer()
+	register(t, "test/resetting_timer", resettingTimer)
+	resettingTimer.Update(time.Second) // must be after register call
+
+	emptyResettingTimer := metrics.NewResettingTimer()
+	register(t, "test/empty_resetting_timer", emptyResettingTimer)
+
+	emptyResettingTimer.Update(time.Second) // no effect because of snapshot below
+	register(t, "test/empty_resetting_timer_snapshot", emptyResettingTimer.Snapshot())
+
+	registerNilMetrics(t, register)
+
+	gatherer := NewGatherer(registry)
+
+	families, err := gatherer.Gather()
+	require.NoError(t, err)
+
+	const expectedString = `
+# TYPE test_counter counter
+test_counter 12345
+# TYPE test_counter_float64 counter
+test_counter_float64 1.1
+# TYPE test_gauge gauge
+test_gauge 23456
+# TYPE test_gauge_float64 gauge
+test_gauge_float64 34567.89
+# TYPE test_histogram summary
+test_histogram{quantile="0.5"} 0
+test_histogram{quantile="0.75"} 0
+test_histogram{quantile="0.95"} 0
+test_histogram{quantile="0.99"} 0
+test_histogram{quantile="0.999"} 0
+test_histogram{quantile="0.9999"} 0
+test_histogram_sum 0
+test_histogram_count 0
+# TYPE test_meter gauge
+test_meter 9.999999e+06
+# TYPE test_resetting_timer summary
+test_resetting_timer{quantile="50"} 1e+09
+test_resetting_timer{quantile="95"} 1e+09
+test_resetting_timer{quantile="99"} 1e+09
+test_resetting_timer_sum 1e+09
+test_resetting_timer_count 1
+# TYPE test_timer summary
+test_timer{quantile="0.5"} 2.25e+07
+test_timer{quantile="0.75"} 4.8e+07
+test_timer{quantile="0.95"} 1.2e+08
+test_timer{quantile="0.99"} 1.2e+08
+test_timer{quantile="0.999"} 1.2e+08
+test_timer{quantile="0.9999"} 1.2e+08
+test_timer_sum 2.3e+08
+test_timer_count 6
+`
+	var (
+		stringReader = strings.NewReader(expectedString)
+		parser       expfmt.TextParser
+	)
+	expectedMetrics, err := parser.TextToMetricFamilies(stringReader)
+	require.NoError(t, err)
```

Comment:
I know I just added this, but I wonder if it would be possible to separate this from this test definition. It looks a lot like we are testing the parsing here, but we are really just generating the expected metrics. Perhaps we could expose this as another helper function `func expectedMetrics(t *testing.T) map[string]*dto.MetricFamily {`

Reply:
After both of the function changes I like `TestGatherer_Gather` a lot more.